cmd/panfs-multipart.go

```diff
@@ -73,68 +106,125 @@ func (fs *PANFSObjects) decodePartFile(name string) (partNumber int, etag string
 }
 
 // Appends parts to an appendFile sequentially.
-func (fs *PANFSObjects) backgroundAppend(ctx context.Context, bucket, object, uploadID string) {
-	fs.appendFileMapMu.Lock()
+func (fs *PANFSObjects) backgroundAppend(ctx context.Context, bucket, object, uploadID string, needFlock bool) {
 	logger.GetReqInfo(ctx).AppendTags("uploadID", uploadID)
 	bucketPath, err := fs.getBucketPanFSPath(ctx, bucket)
 	if err != nil {
 		logger.LogIf(ctx, err, logger.Application)
 		return
 	}
-
+	fs.appendFileMapMu.Lock()
 	file := fs.appendFileMap[uploadID]
 	if file == nil {
-		file = &panfsAppendFile{
-			filePath: pathJoin(bucketPath, panfsS3TmpDir, fs.nodeDataSerial, bgAppendsDirName, fmt.Sprintf("%s.%s", uploadID, mustGetUUID())),
+		file = fs.getPanFSMultipartAppendFile(bucketPath, object, uploadID)
+		if file == nil {
+			fs.appendFileMapMu.Unlock()
+			return
 		}
 		fs.appendFileMap[uploadID] = file
 	}
 	fs.appendFileMapMu.Unlock()
 
-	file.Lock()
-	defer file.Unlock()
+	// No need to take a lock when bgAppend called from CompleteMultipartUpload which has been already taken the lock
+	if needFlock {
+		if !file.flock.TryLock() {
+			return
+		}
+		defer file.flock.Unlock()
+	}
 
-	// Since we append sequentially nextPartNumber will always be len(file.parts)+1
-	nextPartNumber := len(file.parts) + 1
-	uploadIDDir := fs.getUploadIDDir(bucketPath, bucket, object, uploadID)
+	uploadIDDir := fs.getUploadIDDir(bucketPath, object, uploadID)
 
-	entries, err := readDir(uploadIDDir)
+	// Read initial mod time of upload dir
+	prevUploadIDDirStat, err := fsStatDir(ctx, uploadIDDir)
 	if err != nil {
-		logger.GetReqInfo(ctx).AppendTags("uploadIDDir", uploadIDDir)
 		logger.LogIf(ctx, err)
 		return
 	}
-	sort.Strings(entries)
+	fsParts := fs.readMPartUploadParts(bucketPath, object, uploadID)
+	// Since we append sequentially nextPartNumber will always be len(fsParts.Appended)+1
+	initialAppendedLen := len(fsParts.Parts)
```

Comment:
oh, I see that size of part is encoded in file name, so all this magic above is redundant.

Reply:
bgAppend is completely rely on the info from mpart.json file and in case that you described multipart uplodd will be broken and even complete multipart won't detect the issue there. Probably the easiest way (but not ideal) is to add validation of size of the appended file as currently, complete multipart is checking validity of appended file by comparing uploaded parts (from multipart dir) and list of required parts (sent as argument by client) but does not check appended file size